File: apps/flutter-prototype/lib/core/notification_recovery.dart
import 'models.dart';
import 'recovery.dart';

/// Body-free notification routing and durable OS effects in the chat generation.
class RecoveryNotificationJournal {
  RecoveryNotificationJournal([Object? raw]) {
    if (raw == null) {
      _effects['migration'] = {'key': 'migration', 'kind': 'CANCEL_ALL'};
      return;
    }
    if (raw is! Map ||
        raw['version'] != 1 ||
        raw['routes'] is! List ||
        raw['effects'] is! List ||
        raw.keys.any((k) => !{'version', 'routes', 'effects'}.contains(k))) {
      _fail();
    }
    final value = raw;
    for (final r in value['routes'] as List) {
      if (r is! Map ||
          r.length != 5 ||
          r.keys.any(
            (k) => !{
              'messageId',
              'conversationId',
              'objectVersion',
              'accessVersion',
              'streamEpoch',
            }.contains(k),
          ) ||
          !_id(r['messageId']) ||
          !_id(r['conversationId']) ||
          !_id(r['streamEpoch']) ||
          _routes.containsKey(r['messageId'])) {
        _fail();
      }
      recoveryDecimal(r['objectVersion'], allowZero: false);
      recoveryDecimal(r['accessVersion'], allowZero: false);
      _routes[r['messageId'] as String] = Map<String, dynamic>.from(r);
    }
    for (final e in value['effects'] as List) {
      if (e is! Map ||
          !_id(e['key']) ||
          _effects.containsKey(e['key']) ||
          !{'SHOW', 'CANCEL', 'CANCEL_ALL'}.contains(e['kind']) ||
          e.keys.any((k) => !{'key', 'kind', 'messageId'}.contains(k)) ||
          (e['kind'] != 'CANCEL_ALL' && !_id(e['messageId'])) ||
          (e['kind'] == 'SHOW' && !_routes.containsKey(e['messageId']))) {
        _fail();
      }
      _effects[e['key'] as String] = Map<String, dynamic>.from(e);
    }
  }
  final _routes = <String, Json>{}, _effects = <String, Json>{};
  static Never _fail() =>
      throw const RecoveryProtocolError('UNSUPPORTED_SECURE_STATE');
  static bool _id(Object? value) =>
      value is String && value.isNotEmpty && value.length <= 256;
  bool _matches(RecoveryState state, Json r) {
    final message = state.messages[r['messageId']],
        access = state.access[r['conversationId']];
    return state.context.streamEpoch == r['streamEpoch'] &&
        message?.state == MessageRecoveryState.normal &&
        message?.conversationId == r['conversationId'] &&
        message?.objectVersion == r['objectVersion'] &&
        access?.readAllowed == true &&
        access?.accessVersion == r['accessVersion'] &&
        !state.rebuild.contains(r['conversationId']);
  }

  void reconcile(
    RecoveryState state, [
    List<String> liveMessageIds = const [],
  ]) {
    for (final id in _routes.keys.toList()) {
      if (!_matches(state, _routes[id]!)) {
        _routes.remove(id);
        _effects.remove('show:$id');
        _effects['cancel:$id'] = {
          'key': 'cancel:$id',
          'kind': 'CANCEL',
          'messageId': id,
        };
      }
    }
    for (final id in liveMessageIds) {
      if (_routes.containsKey(id)) continue;
      final message = state.messages[id],
          access = message == null
              ? null
              : state.access[message.conversationId];
      if (message?.state != MessageRecoveryState.normal ||
          access?.readAllowed != true ||
          state.rebuild.contains(message!.conversationId)) {
        continue;
      }
      _routes[id] = {
        'messageId': id,
        'conversationId': message.conversationId,
        'objectVersion': message.objectVersion,
        'accessVersion': access!.accessVersion,
        'streamEpoch': state.context.streamEpoch,
      };
      _effects['show:$id'] = {
        'key': 'show:$id',
        'kind': 'SHOW',
        'messageId': id,
      };
    }
  }

  bool allows(RecoveryState state, String id) =>
      state.phase == RecoveryPhase.onlineSafe &&
      _routes.containsKey(id) &&
      _matches(state, _routes[id]!);
  void acknowledge(String key) => _effects.remove(key);
  Json image() => {
    'version': 1,
    'routes': [for (final r in _routes.values) Map<String, dynamic>.of(r)],
    'effects': [for (final e in _effects.values) Map<String, dynamic>.of(e)],
  };
}

File: packages/domain-ts/src/notificationRecovery.ts
import {recoveryMessageVisible,recoveryDecimal,RecoveryProtocolError,type RecoveryState} from './recovery'
export interface NotificationRouteProof {messageId:string;conversationId:string;objectVersion:string;accessVersion:string;streamEpoch:string}
export interface NotificationEffect {key:string;kind:'SHOW'|'CANCEL'|'CANCEL_ALL';messageId?:string}
export interface NotificationJournalImage {version:1;routes:NotificationRouteProof[];effects:NotificationEffect[]}
const fail=():never=>{throw new RecoveryProtocolError('UNSUPPORTED_SECURE_STATE')}
const id=(v:unknown):v is string=>typeof v==='string' && v.length>0 && v.length<=256
/** Body-free routes and OS intents share the chat generation; OS effects run only after commit. */
export class RecoveryNotificationJournal {
  private routes=new Map<string,NotificationRouteProof>()
  private effects=new Map<string,NotificationEffect>()
  constructor(raw?: unknown) {
    if(raw===undefined) {this.effects.set('migration',{key:'migration',kind:'CANCEL_ALL'});return}
    if(!raw || typeof raw!=='object' || Array.isArray(raw)) return fail()
    const value=raw as Record<string,unknown>
    if(Object.keys(value).some(k=>!['version','routes','effects'].includes(k)) || value.version!==1 || !Array.isArray(value.routes) || !Array.isArray(value.effects)) return fail()
    for(const r of value.routes) {
      if(!r || typeof r!=='object' || Object.keys(r).sort().join(',')!=='accessVersion,conversationId,messageId,objectVersion,streamEpoch'
        || !id(r.messageId) || !id(r.conversationId) || !id(r.streamEpoch) || this.routes.has(r.messageId)) return fail()
      recoveryDecimal(r.objectVersion,false);recoveryDecimal(r.accessVersion,false)
      this.routes.set(r.messageId,{...r})
    }
    for(const e of value.effects) {
      if(!e || typeof e!=='object' || !id(e.key) || this.effects.has(e.key) || !['SHOW','CANCEL','CANCEL_ALL'].includes(e.kind)
        || Object.keys(e).some(k=>!['key','kind','messageId'].includes(k)) || (e.kind!=='CANCEL_ALL' && !id(e.messageId))
        || (e.kind==='SHOW' && !this.routes.has(e.messageId))) return fail()
      this.effects.set(e.key,{...e})
    }
  }
  reconcile(state: RecoveryState, liveMessageIds: readonly string[]=[]): void {
    const matches=(r:NotificationRouteProof)=>state.context.streamEpoch===r.streamEpoch
      && state.messages.get(r.messageId)?.state==='NORMAL' && state.messages.get(r.messageId)?.conversationId===r.conversationId
      && state.messages.get(r.messageId)?.objectVersion===r.objectVersion && state.access.get(r.conversationId)?.readAllowed===true
      && state.access.get(r.conversationId)?.accessVersion===r.accessVersion && !state.rebuild.has(r.conversationId)
    for(const [id,r] of this.routes) if(!matches(r)) {
      this.routes.delete(id);this.effects.delete(`show:${id}`)
      this.effects.set(`cancel:${id}`,{key:`cancel:${id}`,kind:'CANCEL',messageId:id})
    }
    for(const id of liveMessageIds) {
      if(this.routes.has(id)) continue
      const message=state.messages.get(id),access=message && state.access.get(message.conversationId)
      if(message?.state!=='NORMAL' || !access?.readAllowed || state.rebuild.has(message.conversationId)) continue
      this.routes.set(id,{messageId:id,conversationId:message.conversationId,objectVersion:message.objectVersion,accessVersion:access.accessVersion,streamEpoch:state.context.streamEpoch})
      this.effects.set(`show:${id}`,{key:`show:${id}`,kind:'SHOW',messageId:id})
    }
  }
  allows(state:RecoveryState,id:string):boolean {
    const route=this.routes.get(id)
    return Boolean(route && recoveryMessageVisible(state,id) && state.context.streamEpoch===route.streamEpoch
      && state.messages.get(id)?.conversationId===route.conversationId && state.messages.get(id)?.objectVersion===route.objectVersion && state.access.get(route.conversationId)?.accessVersion===route.accessVersion)
  }
  acknowledge(key:string):void {this.effects.delete(key)}
  image():NotificationJournalImage {return {version:1,routes:[...this.routes.values()].map(r=>({...r})),effects:[...this.effects.values()].map(e=>({...e}))}}
}
